Menus/InventoryMenuGUI.cpp

```diff
@@ -0,0 +1,1237 @@
+		m_GUIInventoryItemButtons.clear();
+		m_GUIInventoryItemButtons.reserve(c_FullViewPageItemLimit);
+
+		m_GUIPopupBox = nullptr;
+		m_GUIPopupText = nullptr;
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	int InventoryMenuGUI::Create(Controller *activityPlayerController, Actor *inventoryActor, MenuMode menuMode) {
+		RTEAssert(activityPlayerController, "No controller sent to InventoryMenuGUI on creation!");
+		RTEAssert(c_ItemsPerRow % 2 == 1, "Don't you dare use an even number of items per inventory row, you filthy animal!");
+
+		if (!m_SmallFont) { m_SmallFont = g_FrameMan.GetSmallFont(); }
+		if (!m_LargeFont) { m_LargeFont = g_FrameMan.GetLargeFont(); }
+
+		m_MenuController = activityPlayerController;
+		SetInventoryActor(inventoryActor);
+		m_MenuMode = menuMode;
+
+		if (m_MenuMode == MenuMode::Carousel && !CarouselModeReadyForUse()) {
+			return SetupCarouselMode();
+		} else if ((m_MenuMode == MenuMode::Full || m_MenuMode == MenuMode::Transfer) && !FullOrTransferModeReadyForUse()) {
+			return SetupFullOrTransferMode();
+		}
+
+		return 0;
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	int InventoryMenuGUI::SetupCarouselMode() {
+		for (std::unique_ptr<CarouselItemBox> &carouselItemBox : m_CarouselItemBoxes) {
+			carouselItemBox = std::make_unique<CarouselItemBox>();
+			carouselItemBox->IsForEquippedItems = false;
+		}
+		std::size_t equippedItemIndex = c_ItemsPerRow / 2;
+		Vector currentBoxSize = c_CarouselBoxMaxSize;
+		int carouselBitmapWidth = 0;
+		for (int i = equippedItemIndex; i >= 0; i--) {
+			m_CarouselItemBoxes.at(i)->FullSize = currentBoxSize;
+			m_CarouselItemBoxes.at(i)->IconCenterPosition.SetY(c_CarouselBoxMaxSize.GetY() / 2 + static_cast<float>(m_SmallFont->GetFontHeight() / 2));
+			carouselBitmapWidth += currentBoxSize.GetFloorIntX();
+			if (i != equippedItemIndex) {
+				m_CarouselItemBoxes.at((equippedItemIndex * 2) - i)->FullSize = currentBoxSize;
+				m_CarouselItemBoxes.at((equippedItemIndex * 2) - i)->IconCenterPosition.SetY(c_CarouselBoxMaxSize.GetY() / 2 + static_cast<float>(m_SmallFont->GetFontHeight() / 2));
+				carouselBitmapWidth += currentBoxSize.GetFloorIntX();
+			} else {
+				m_CarouselItemBoxes.at(i)->IsForEquippedItems = true;
+			}
+			currentBoxSize -= c_CarouselBoxSizeStep;
+		}
+
+		m_CarouselExitingItemBox = std::make_unique<CarouselItemBox>();
+		m_CarouselExitingItemBox->IsForEquippedItems = false;
+		m_CarouselExitingItemBox->IconCenterPosition.SetY(c_CarouselBoxMaxSize.GetY() / 2 + static_cast<float>(m_SmallFont->GetFontHeight() / 2));
+		m_CarouselExitingItemBox->FullSize = c_CarouselBoxMinSize - c_CarouselBoxSizeStep;
+
+		m_CarouselBitmap = std::unique_ptr<BITMAP>(create_bitmap_ex(8, carouselBitmapWidth, c_CarouselBoxMaxSize.GetFloorIntY() + m_SmallFont->GetFontHeight() / 2));
```

Comment:
Shit, good catch. I could specify a custom deleter for this instead though, no?
Also, why do you say make Destroy virtual?

Reply:
Just don't introduce cleverassness that is absolutely unnecessary.
You wouldn't know that your custom deleter is working properly without actually testing for leakage.

And i'm saying make the destructor virtual, not Destroy.
If you have virtuals in the class you need a virtual destructor. It's not gonna do any change as is, but if you'd to inherit from this class you'd want the child class' dtor to override the base one so they cascade and everything destructs proper.
But you're not planning this to be a base class so you don't actually need any virtuals, so when you yeet them just make a regular non-virtual one.

Reply:
Gonna move this one to pms. There's some stuff that needs to be sorted out.

Reply:
Hmmm yeah, I wasn't thinking about this stuff properly at all I guess. I wanted to get rid of the explicit destructor and Destroy and just rely on Reset, but that causes this issue (and the same in PieMenuGUI).

I'm not a huge fan of having to add an explicit destructor and a boilerplate Destroy just for this, the custom deleter option seems more elegant to me, but you're right that it's untested. I think figuring out if this can work could be good for future though, move us away from having to have all that explicity destruction handling by moving to properly managed memory.

Reply:
If SDL fails then we already have a "managed" bitmap type we can use, AllegroBitmap.
We just need to put the effort to swap all raw BITMAPs to those instead.

Reply:
That's true I guess. I dunno if it'd be good to make that swap or not though.
But yeah, are you very against the custom deleter option? If you are, I'll get rid of it and add the destruction boilerplate, I think I've explained my thoughts here, but I'm ultimately okay either way.

Reply:
I guess I am. I don't mind the boilerplate too much.